features/keychain/module/multi-seed-keychain.js

```diff
@@ -0,0 +1,145 @@
+import assert from 'minimalistic-assert'
+import { Keychain } from './keychain'
+import { getSeedId } from './crypto/seed-id'
+
+const MODULE_ID = 'multiSeedKeychain'
+
+class MultiSeedKeychain {
+  #keychains = Object.create(null)
+  #primarySeedId
+  #legacyPrivToPub
+  #logger
+
+  constructor({ legacyPrivToPub = Object.create(null), logger }) {
+    this.#legacyPrivToPub = legacyPrivToPub
+    this.#logger = logger
+    this.sodium = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ seedIdentifier, ...rest }) =>
+            this.#getKeychainForSeed(seedIdentifier).sodium[key](rest),
+      }
+    )
+
+    this.secp256k1 = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ seedIdentifier, ...rest }) =>
+            this.#getKeychainForSeed(seedIdentifier).secp256k1[key](rest),
+      }
+    )
+
+    this.ed25519 = new Proxy(
+      {},
+      {
+        get:
+          (target, key) =>
+          async ({ seedIdentifier, ...rest }) =>
+            this.#getKeychainForSeed(seedIdentifier).ed25519[key](rest),
+      }
+    )
+  }
+
+  #getKeychainForSeed = (seedIdentifier) => {
+    assert(this.#keychains, 'keychain is locked')
+    assert(typeof seedIdentifier === 'string', 'expected seedIdentifier')
+
+    const seedIdentifierHex = seedIdentifier.toString('hex')
+    const keychain = this.#keychains[seedIdentifierHex]
+    assert(keychain, `keychain not found for seed id: ${seedIdentifierHex}`)
+    return keychain
+  }
+
+  lock = () => {
+    this.#keychains = null
+    this.#primarySeedId = null
+  }
+
+  #initKeychain = (seed) => {
+    assert(Buffer.isBuffer(seed) && seed.length === 64, 'seed must be a 64 byte buffer')
+
+    if (!this.#keychains) this.#keychains = Object.create(null)
+
+    const seedId = getSeedId(seed)
+    if (this.#keychains[seedId]) throw new Error(`already have seed with id: ${seedId}`)
+
+    const keychain = new Keychain({
+      legacyPrivToPub: this.#legacyPrivToPub,
+      logger: this.#logger,
+    })
+
+    keychain.unlock({ seed })
+
+    this.#keychains[seedId] = keychain
+    return seedId
+  }
+
+  setPrimarySeed = (seed) => {
+    this.#primarySeedId = this.#initKeychain(seed)
+    return this.#primarySeedId
+  }
+
+  addSeed = (seed) => {
+    assert(this.#keychains, 'keychain is locked')
+    assert(this.#primarySeedId, 'primary seed not set')
+    return this.#initKeychain(seed)
+  }
```

Comment:
does the keychain need to know the difference between a primary seed and extra seeds if the public api is extended with the seed id (as you've already done further done with `exportKey` f.i.)

Reply:
i was thinking some callers wouldn't know the primary seed id, but we can refactor those to be explicit